sync/src/tests/sync_shared_state.rs

```diff
@@ -0,0 +1,125 @@
+use crate::tests::util::{build_chain, inherit_block};
+use crate::types::BlockStatus;
+use crate::SyncSharedState;
+use ckb_chain::chain::ChainService;
+use ckb_core::block::BlockBuilder;
+use ckb_core::header::HeaderBuilder;
+use ckb_core::Capacity;
+use ckb_db::MemoryKeyValueDB;
+use ckb_network::PeerIndex;
+use ckb_notify::NotifyService;
+use ckb_shared::shared::SharedBuilder;
+use ckb_store::ChainStore;
+use ckb_test_chain_utils::always_success_cellbase;
+use std::sync::Arc;
+
+#[test]
+fn test_insert_new_block() {
+    let (shared, chain) = build_chain(2);
+    let new_block = {
+        let tip_number = shared.tip_header().number();
+        let next_block = inherit_block(shared.shared(), tip_number).build();
+        Arc::new(next_block)
+    };
+
+    assert_eq!(
+        shared
+            .insert_new_block(&chain, PeerIndex::new(1), Arc::clone(&new_block))
+            .expect("insert valid block"),
+        true,
+    );
+    assert_eq!(
+        shared
+            .insert_new_block(&chain, PeerIndex::new(1), Arc::clone(&new_block))
+            .expect("insert duplicated valid block"),
+        false,
+    );
+}
+
+#[test]
+fn test_insert_invalid_block() {
+    let (shared, chain) = build_chain(2);
+    let invalid_block = {
+        let tip_number = shared.tip_header().number();
+        let invalid_cellbase = always_success_cellbase(tip_number, Capacity::zero());
+        let next_block = inherit_block(shared.shared(), tip_number)
+            .transaction(invalid_cellbase)
+            .build();
+        Arc::new(next_block)
+    };
+
+    assert!(shared
+        .insert_new_block(&chain, PeerIndex::new(1), Arc::clone(&invalid_block))
+        .is_err(),);
```

Comment:
`is_err` is obscure, assert the exact err type.

Reply:
It is `FailureError`. assert it is tedious...